c/tskit/tables.c

```diff
@@ -8105,6 +8105,285 @@ tsk_table_collection_subset(
     return ret;
 }
 
+static int
+tsk_check_subset_equality(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *node_mapping, tsk_size_t other_num_nodes, tsk_size_t num_shared_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, i;
+    tsk_id_t *self_nodes = NULL;
+    tsk_id_t *other_nodes = NULL;
+    tsk_table_collection_t self_copy;
+    tsk_table_collection_t other_copy;
+
+    self_nodes = malloc(num_shared_nodes * sizeof(*self_nodes));
+    other_nodes = malloc(num_shared_nodes * sizeof(*other_nodes));
+    if (self_nodes == NULL || other_nodes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    ret = tsk_table_collection_init(&self_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_init(&other_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    i = 0;
+    for (k = 0; k < (tsk_id_t) other_num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            self_nodes[i] = node_mapping[k];
+            other_nodes[i] = k;
+            i++;
+        }
+    }
+
+    // TODO: strict sort before checking equality
+    ret = tsk_table_collection_copy(self, &self_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_copy(other, &other_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&other_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&self_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&self_copy, self_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&other_copy, other_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = !tsk_table_collection_equals(&self_copy, &other_copy);
+    if (ret != 0) {
+        ret = TSK_ERR_DIFF_HISTORIES;
+        goto out;
+    }
+
+out:
+    tsk_table_collection_free(&self_copy);
+    tsk_table_collection_free(&other_copy);
+    tsk_safe_free(other_nodes);
+    tsk_safe_free(self_nodes);
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *node_mapping, tsk_size_t num_nodes, tsk_flags_t options)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_ind, new_pop, new_parent, new_child;
+    tsk_size_t num_shared_nodes = 0;
+    tsk_node_t node;
+    tsk_individual_t ind;
+    tsk_population_t pop;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    tsk_id_t *node_map = NULL;
```

Comment:
It's a bit unfortunate to have ``node_map`` and ``node_mapping``. It's confusing to pull apart what they are in the code. How about calling the parameter ``other_node_mapping``?